Testing-the-engine.md

```diff
@@ -0,0 +1,302 @@
+of creating unit tests in Xcode projects. Since iOS has x86 simulators and
+since we can build x86 engines, we can execute the XCTests directly on macOS
+in a headless simulator using the real iOS SDK.
+
+#### OCMock
+
+[OCMock](https://ocmock.org) is a standard iOS testing library used to mock
+dependencies needed to construct and test interactions with your under-test
+production class.
+
+It's best practice to test only one real production class per test and
+mock all other dependencies with OCMock.
+
+The OCMock library is an available test dependency when writing XCTests.
+
+### End-to-end tests
+
+End-to-end tests exercise the entire iOS embedding with the C++ engine on
+a headless iOS simulator. It's an integration test ensuring that
+the engine as a whole on iOS is functioning correctly.
+
+The project containing the iOS end-to-end engine test is at
+https://github.com/flutter/engine/tree/master/testing/scenario_app/ios.
+
+This test project is build similarly to a normal debug Flutter app. With the
+Dart code bundled in JIT mode and the iOS part brought into Xcode with a
+`.framework` dependency on the prebuilt local engine. It's then installed and
+executed on a simulator via Xcode.
+
+Unlike a normal Flutter app, the Flutter framework on the Dart side is a
+lightweight fake at https://github.com/flutter/engine/tree/master/testing/scenario_app/lib
+that implements some of the basic functionalities of `dart:ui` Window rather
+than using the real Flutter framework at `flutter/flutter`.
+
+The end-to-end test can be executed by running:
+
+    `testing/scenario_app/build_and_run_ios_tests.sh`
+
+Additional end-to-end instrumented tests can be added to https://github.com/flutter/engine/tree/master/testing/scenario_app/ios/Scenarios/ScenariosTests.
+
+If supporting logic are needed for the test case, they can be added to the
+Android app under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/ios/Scenarios/Scenarios
+or to the fake Flutter framework under-test in https://github.com/flutter/engine/tree/master/testing/scenario_app/lib.
+
+As best practice, favor adding unit tests if possible since end-to-end tests
+are, by nature, non-hermetic, slow and flaky.
+
+End-to-end tests on iOS are executed during pre-submit on our CI system when
+submitting PRs to the `flutter/engine` repository.
+
+## Dart - dart:ui
+
+If you're editing `.dart` files in https://github.com/flutter/engine/tree/master/lib/ui,
+you're working on the 'dart:ui' package which is the interface between
+the C++ engine and the Dart Flutter framework.
+
+### Unit tests
+
+Dart classes in https://github.com/flutter/engine/tree/master/lib/ui have matching
+unit tests at https://github.com/flutter/engine/tree/master/testing/dart.
+
+When editing production files in the 'dart:ui' package, add to or create a
+test file in `testing/dart`.
+
+To run the Dart unit tests, use the script:
+
+   `testing/run_tests.py --type=dart`
```

Comment:
Github doesn't render these backquotes (because it is only a three space indent), but you might want to put then into a block backquote just so it renders it the same as the other commands.